pkg/runtime/controller.go

```diff
@@ -0,0 +1,420 @@
+
+	// new handler factory is created for every config change.
+	hb := c.createHandlerFactory(c.templateInfo, c.eval, c.attrDescFinder, c.adapterInfo)
+
+	// new handler table is created for every config change. It uses handler factory
+	// to create new handlers.
+	ht := newHandlerTable(instanceConfig, handlerConfig,
+		func(handler *cpb.Handler, instances []*cpb.Instance) (adapter.Handler, error) {
+			return hb.Build(handler, instances, newEnv(handler.Name, c.handlerPool))
+		},
+	)
+
+	// current consistent view of the rules keyed by Namespace and then Name.
+	// ht (handlerTable) keeps track of handler-instance association.
+	ruleConfig := c.processRules(handlerConfig, instanceConfig, ht)
+
+	// Initialize handlers that are used in the configuration.
+	// Some handlers may not initialize due to errors.
+	ht.Initialize(c.table)
+
+	// Combine rules with the handler table.
+	// Actions referring to handlers in error, will be purged.
+	combineRulesHandlers(ruleConfig, ht.table)
+
+	// create rules that are in the format that resolver needs.
+	rules, nrules := convertToRuntimeRules(ruleConfig)
+
+	// Create new resolver and cleanup the old resolver.
+	c.resolverID++
+	resolver := newResolver(c.eval, c.identityAttribute, c.defaultConfigNamespace, rules, c.resolverID)
+	c.dispatcher.ChangeResolver(resolver)
+
+	// copy old for deletion.
+	oldTable := c.table
+	oldResolver := c.resolver
+
+	// set new
+	c.table = ht.table
+	c.resolver = resolver
+
+	glog.Infof("Published snapshot[%d] with %d rules, %d handlers", resolver.ID, nrules, len(c.table))
+
+	// synchronous call to cleanup.
+	err := cleanupResolver(oldResolver, oldTable, maxCleanupDuration)
+	if err != nil {
+		glog.Warningf("Unable to perform cleanup: %v", err)
+	}
+
+}
+
+// maxCleanupDuration is the maximum amount of time cleanup operation will wait
+// before resolver ref count does to 0. It will return after this duration without
+// calling Close() on handlers.
+var maxCleanupDuration = 10 * time.Second
+
+var watchFlushDuration = time.Second
+
+// watchChanges watches for changes and publishes a new resolver.
+// watchChanges is started in a goroutine.
+func watchChanges(wch <-chan store.Event, applyEvents applyEventsFn) {
+	// consume changes and apply them to data indefinitely
+	var timeChan <-chan time.Time
+	var timer *time.Timer
+	events := make([]*store.Event, 0, maxEvents)
+
+	for {
+		select {
+		case ev := <-wch:
+			if len(events) == 0 {
+				timer = time.NewTimer(watchFlushDuration)
+				timeChan = timer.C
+			}
+			events = append(events, &ev)
+		case <-timeChan:
+			timer.Stop()
+			timeChan = nil
+			glog.Infof("Publishing %d events", len(events))
+			applyEvents(events)
+			events = events[:0]
+		}
+	}
+}
+
+// applyEvents applies given events to config state and then publishes a snapshot.
+func (c *Controller) applyEvents(events []*store.Event) {
+	for _, ev := range events {
+		switch ev.Type {
+		case store.Update:
+			c.configState[ev.Key] = ev.Value
```

Comment:
I don't think so. applyEvents is called from watchChanges, within a synchronous for-loop. No other applyEvents are called at the same time.
(and the use of configState is within publishSnapshot, which is only called from applyEvents and before watchChanges goroutine starts).